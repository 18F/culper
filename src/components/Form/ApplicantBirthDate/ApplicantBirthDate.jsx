import React from 'react'
import ValidationElement from '../validationElement'
import DateControl from '../DateControl'

export default class ApplicantBirthDate extends ValidationElement {
  constructor (props) {
    super(props)

    this.state = {
      value: props.value,
      estimated: props.estimated
    }
  }

  render () {
    return (
      <div>
        <DateControl name={this.props.name}
                     value={this.state.value}
                     estimated={this.state.estimated}
                     />
      </div>
    )
  }
}
