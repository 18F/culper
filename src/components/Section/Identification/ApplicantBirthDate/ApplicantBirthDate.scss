.birthdate {
  .datecontrol {
    display: inline-block;
  }

  .age-warning {
    display: block;

    label {
      width: 30rem;
    }
  }
}
