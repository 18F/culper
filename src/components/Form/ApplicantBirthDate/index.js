import ApplicantBirthDate from './ApplicantBirthDate'
export default ApplicantBirthDate
