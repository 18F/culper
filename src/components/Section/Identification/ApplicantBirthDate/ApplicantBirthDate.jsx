import React from 'react'
import { i18n } from '../../../../config'
import schema from '../../../../schema'
import validate from '../../../../validators'
import SubsectionElement from '../../SubsectionElement'
import { Field, DateControl, Show, Checkbox } from '../../../Form'

export default class ApplicantBirthDate extends SubsectionElement {
  constructor (props) {
    super(props)

    this.state = {
      uid: `${this.props.name}-${super.guid()}`,
      errors: [],
      needsConfirmation: false
    }

    this.update = this.update.bind(this)
    this.updateDate = this.updateDate.bind(this)
    this.updateConfirmed = this.updateConfirmed.bind(this)
    this.handleError = this.handleError.bind(this)
    this.confirmationError = this.confirmationError.bind(this)
  }

  /**
   * Handle the change event.
   */
  update (queue) {
    this.props.onUpdate({
      Date: this.props.Date,
      Confirmed: this.props.Confirmed,
      ...queue
    })
  }

  updateDate (values) {
    this.update({
      Date: values,
      Confirmed: { value: '', checked: false }
    })
  }

  updateConfirmed (values) {
    this.update({
      Confirmed: values
    })
  }

  handleError (value, arr) {
    let local = [...arr]

    const hasMinMaxError = local.some(x => x.valid === false && (x.code === 'date.min' || x.code === 'date.max'))
    const birthdateValid = this.props.Confirmed && this.props.Confirmed.checked === true
          ? true
          : hasMinMaxError ? false : null
    local.push({
      code: 'birthdate.age',
      valid: birthdateValid,
      uid: this.state.uid
    })
    local = local.filter(x => x.code !== 'date.min' && x.code !== 'date.max')

    // Store the errors
    this.setState({ errors: local, needsConfirmation: hasMinMaxError })

    // Take the original and concatenate our new error values to it
    return super.handleError(value, local)
  }

  confirmationError (value, arr) {
    let local = [...this.state.errors]
    local = local.filter(x => x.code !== 'birthdate.age')
    local.push({
      code: 'birthdate.age',
      valid: value,
      uid: this.state.uid
    })

    // Store the errors
    this.setState({ errors: local })

    // Take the original and concatenate our new error values to it
    return super.handleError(value, local)
  }

  render () {
    const klass = `section-content birthdate ${this.props.className || ''}`.trim()
    return (
      <div className={klass} {...super.dataAttributes(this.props)}>
        <Field title={i18n.t('identification.birthdate.title')}
               titleSize="h2"
               help="identification.birthdate.help"
               scrollIntoView={this.props.scrollIntoView}>
          <DateControl name={this.props.name}
                       {...this.props.Date}
                       relationship="Self"
                       overrideError={(this.props.Confirmed || {}).checked}
                       onUpdate={this.updateDate}
                       onError={this.handleError}
                       required={this.props.required}
                       />
          <Show when={this.state.needsConfirmation}>
            <Checkbox {...this.props.Confirmed}
                      name="Confirmed"
                      label={i18n.t('identification.birthdate.confirmation')}
                      value="Confirmed"
                      className="age-warning"
                      onUpdate={this.updateConfirmed}
                      onError={this.confirmationError}
                      required={this.state.needsConfirmation && this.props.required}
                      />
          </Show>
        </Field>
      </div>
    )
  }
}

ApplicantBirthDate.defaultProps = {
  Date: {},
  Confirmed: {},
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr },
  section: 'identification',
  subsection: 'birthdate',
  dispatch: () => {},
  validator: (data) => {
    return validate(schema('identification.birthdate', data))
  }
}
