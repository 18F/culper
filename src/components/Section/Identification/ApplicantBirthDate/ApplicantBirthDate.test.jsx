import React from 'react'
import { shallow, mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { ApplicantBirthDate } from './ApplicantBirthDate'

// give a fake GUID so the field IDs don't differ between snapshots
jest.mock('components/Form/ValidationElement/index')

describe('The ApplicantBirthDate component', () => {
  const mockStore = configureMockStore()

  const mountComponentWithStore = (props = {}, defaultState = {}) => {
    const store = mockStore({ ...defaultState })
    return mount(
      <Provider store={store}>
        <ApplicantBirthDate {...props} />
      </Provider>
    )
  }

  it('renders without errors', () => {
    const component = shallow(<ApplicantBirthDate />)

    expect(component.exists()).toBe(true)
    expect(component).toMatchSnapshot()
  })

  it('displays no error by default', () => {
    const component = mountComponentWithStore()

    component.find('.month input').simulate('change')
    expect(component.find('.usa-input-error-label').length).toEqual(0)
  })

  it('implements an onUpdate handler', () => {
    let updates = 0
    const testProps = {
      onUpdate: () => {
        updates += 1
      },
    }

    const component = mountComponentWithStore(testProps)
    component.find('.month input').simulate('change')
    component.find('.day input').simulate('change')
    component.find('.year input').simulate('change')
    expect(updates).toBe(3)
  })

  it('displays age confirmation', () => {
    const testProps = {
      Date: {
        month: '01',
        day: '01',
        year: '1700',
      },
    }

    const component = mountComponentWithStore(testProps)
    expect(component.find('.age-warning').length).toBe(1)
  })
})
