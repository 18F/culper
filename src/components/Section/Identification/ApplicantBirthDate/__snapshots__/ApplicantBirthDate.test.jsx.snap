// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ApplicantBirthDate component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="section-content birthdate"
    data-section="identification"
    data-subsection="birthdate"
  >
    <h1
      className="section-header"
    >
      Date of birth
    </h1>
    <Field
      adjustFor=""
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.birthdate.help"
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Provide your date of birth"
      titleSize="h4"
      validate={true}
    >
      <Connect(DateControl)
        name="birthdate"
        onError={[Function]}
        onUpdate={[Function]}
        relationship="Self"
      />
      <Show
        when={false}
      >
        <Checkbox
          checked={false}
          className="age-warning"
          error={false}
          focus={false}
          label="I confirm this date is correct"
          name="Confirmed"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          valid={false}
          value="Confirmed"
        />
      </Show>
    </Field>
  </div>,
  "nodes": Array [
    <div
      className="section-content birthdate"
      data-section="identification"
      data-subsection="birthdate"
    >
      <h1
        className="section-header"
      >
        Date of birth
      </h1>
      <Field
        adjustFor=""
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.birthdate.help"
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Provide your date of birth"
        titleSize="h4"
        validate={true}
      >
        <Connect(DateControl)
          name="birthdate"
          onError={[Function]}
          onUpdate={[Function]}
          relationship="Self"
        />
        <Show
          when={false}
        >
          <Checkbox
            checked={false}
            className="age-warning"
            error={false}
            focus={false}
            label="I confirm this date is correct"
            name="Confirmed"
            onError={[Function]}
            onUpdate={[Function]}
            required={false}
            valid={false}
            value="Confirmed"
          />
        </Show>
      </Field>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ApplicantBirthDate
        Confirmed={Object {}}
        Date={Object {}}
        dispatch={[Function]}
        onError={[Function]}
        onUpdate={[Function]}
        validator={[Function]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ApplicantBirthDate {
        "_reactInternalInstance": [Circular],
        "confirmationError": [Function],
        "context": Object {},
        "handleCompletion": [Function],
        "handleError": [Function],
        "props": Object {
          "Confirmed": Object {},
          "Date": Object {},
          "dispatch": [Function],
          "onError": [Function],
          "onUpdate": [Function],
          "validator": [Function],
        },
        "refs": Object {},
        "section": "identification",
        "state": Object {
          "errors": Array [],
          "needsConfirmation": false,
          "uid": "birthdate-MOCK-GUID",
        },
        "store": "Identification",
        "storeKey": "ApplicantBirthDate",
        "subsection": "birthdate",
        "update": [Function],
        "updateConfirmed": [Function],
        "updateDate": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="section-content birthdate"
          data-section="identification"
          data-subsection="birthdate"
        >
          <h1
            className="section-header"
          >
            Date of birth
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.birthdate.help"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your date of birth"
            titleSize="h4"
            validate={true}
          >
            <Connect(DateControl)
              name="birthdate"
              onError={[Function]}
              onUpdate={[Function]}
              relationship="Self"
            />
            <Show
              when={false}
            >
              <Checkbox
                checked={false}
                className="age-warning"
                error={false}
                focus={false}
                label="I confirm this date is correct"
                name="Confirmed"
                onError={[Function]}
                onUpdate={[Function]}
                required={false}
                valid={false}
                value="Confirmed"
              />
            </Show>
          </Field>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="section-content birthdate"
          data-section="identification"
          data-subsection="birthdate"
        >
          <h1
            className="section-header"
          >
            Date of birth
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.birthdate.help"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your date of birth"
            titleSize="h4"
            validate={true}
          >
            <Connect(DateControl)
              name="birthdate"
              onError={[Function]}
              onUpdate={[Function]}
              relationship="Self"
            />
            <Show
              when={false}
            >
              <Checkbox
                checked={false}
                className="age-warning"
                error={false}
                focus={false}
                label="I confirm this date is correct"
                name="Confirmed"
                onError={[Function]}
                onUpdate={[Function]}
                required={false}
                valid={false}
                value="Confirmed"
              />
            </Show>
          </Field>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ApplicantBirthDate
    Confirmed={Object {}}
    Date={Object {}}
    dispatch={[Function]}
    onError={[Function]}
    onUpdate={[Function]}
    validator={[Function]}
  />,
}
`;
